harness/determined/pytorch/deepspeed/_deepspeed_context.py

```diff
@@ -0,0 +1,338 @@
+       model engine.
+    2. Overwrite a deepspeed config file or dictionary with values from Determined's
+       experiment config to ensure consistency in batch size and support hyperparameter tuning.
+    3. Set a custom model parallel configuration that should instantiate a
+       :class:`determined.pytorch.deepspeed.ModelParallelUnit` dataclass.  We automatically set the
+       mpu for data parallel and standard pipeline parallel training.  This should only be needed
+       if there is additional model parallelism outside of DeepSpeed's supported methods.
+    4. Disable data reproducibility checks to allow custom dataloaders.
+    5. Disable automatic gradient aggregation for non-pipeline-parallel training.
+    """
+
+    def __init__(self, *args: Any, **kwargs: Any) -> None:
+        det.TrialContext.__init__(self, *args, **kwargs)
+        pytorch._PyTorchReducerContext.__init__(self, self.distributed._zmq_allgather)
+
+        self._init_device()
+
+        # Track which types we have issued warnings for in to_device().
+        self._to_device_warned_types = set()  # type: Set[Type]
+
+        # DeepSpeed supports mixed precision through NVidia Apex AMP.  ZeRO optimizer requires
+        # Apex AMP and cannot be used with more complex AMP modes.
+        apex_available = importlib.util.find_spec("apex") is not None
+        if not apex_available:
+            logging.warning(
+                "Missing package APEX is required for ZeRO optimizer support through DeepSpeed."
+            )
+
+        # The following attributes are initialized during the lifetime of
+        # a DeepSpeedTrialContext.
+        self.models = []  # type: List[deepspeed.DeepSpeedEngine]
+        self._epoch_len = None  # type: Optional[int]
+
+        self._loss_ids = {}  # type: Dict[torch.Tensor, int]
+        self._last_backward_batch_idx = None  # type: Optional[int]
+        self._current_batch_idx = None  # type: Optional[int]
+
+        self._determined_profiler = None  # type: Optional[profiler.ProfilerAgent]
+        self._mpu = det_ds.get_data_parallel_mpu(self.distributed)  # type: det_ds.ModelParallelUnit
+        self._called_set_mpu = False
+        self._train_micro_batch_size_per_gpu = None  # type: Optional[int]
+        self._num_micro_batches_per_slot = None  # type: Optional[int]
+        self._use_pipeline_parallel = False
+        self._data_repro_checks_disabled = False
+        self._manual_grad_accumulation = False
+
+        self._check_experiment_config_optimizations()
+
+    def _check_experiment_config_optimizations(self) -> None:
+        """
+        Check if the user specified options in optimizations are incompatible with
+        DeepSpeedTrial.
+        """
+        optimizations_config = self.env.experiment_config.get_optimizations_config()
+        self._average_training_metrics = cast(
+            bool, optimizations_config.get("average_training_metrics")
+        )
```

Comment:
instead of a `cast`, can you just do `bool(optimizations_config.get("average_training_metrics", False)`?